---
title: "mm21.08 CellChat"
author: "Shoumit Dey"
date: '2022-07-09'
output: html_document
---


```{r load_libraries, include=FALSE}
#Guide followed: https://htmlpreview.github.io/?https://github.com/sqjin/CellChat/blob/master/tutorial/CellChat-vignette.html
library(CellChat)
library(patchwork)
library(Seurat)
library(stringr)
library(NMF)
library(ggalluvial)
library(wordcloud)
#save all files and folder within your working directory or working directory/your_custom_folder_name/
#update exp_path if you have a your_custom_folder_name within working directory 
exp_path <- paste0(getwd(),"/")
save_within <- paste0(exp_path,"/cellchat/")
```


#FOR FIGURES 6A and S9B
```{r cellchat_comparison_functions}
figures<-c("fig4a","fig4b","fig5a","fig5b")
comparisons<-list()

#comparison direction for cellchat is second 
#versus one. 
comparisons[["fig4a"]]<-"Untd_INF!Untd_d0"
comparisons[["fig4b"]]<-"Untd_d0!UVB_d0"
comparisons[["fig5a"]]<-"UVB_INF!UVB_d0"
comparisons[["fig5b"]]<-"UVB_INF!Untd_INF"


##############################################
#Figures 6A and S9B###########################
##############################################
pdf(paste0(save_within, "information_flow.pdf"), height = 6, width = 6)
  for(figure in 1:length(figures)){
      #check figure number
      print(figures[[figure]])
      #check comparison
      print(comparisons[[figures[[figure]]]])
      #split comparison string
      comp_list<-str_split(comparisons[[figures[[figure]]]], "!")
      #print current comparison
      print(paste0(comp_list[[1]][1]," Versus ", comp_list[[1]][2]))
      
      read_path1 <- paste0(save_within, comp_list[[1]][1], "/")
      read_path2 <- paste0(save_within, comp_list[[1]][2], "/")
      
      cellchat1 <- readRDS(paste0(read_path1,"cellchat_",comp_list[[1]][1],".rds"))
      cellchat2 <- readRDS(paste0(read_path2,"cellchat_",comp_list[[1]][2],".rds"))
      
      object.list <- list()
      object.list[[comp_list[[1]][1]]] <- cellchat1
      object.list[[comp_list[[1]][2]]] <- cellchat2
      
      cellchat <- mergeCellChat(object.list, add.names = names(object.list))
      
      #Differential analysis
      # define a positive dataset, i.e., the dataset with positive fold change against the other dataset
      pos.dataset = comp_list[[1]][2]
      # define a char name used for storing the results of differential expression analysis
      features.name = pos.dataset
      # perform differential expression analysis
      cellchat <- identifyOverExpressedGenes(cellchat, group.dataset = "datasets", pos.dataset = pos.dataset, features.name = features.name, only.pos = FALSE, thresh.pc = 0.1, thresh.fc = 0.1, thresh.p = 1)
  
      #Compare the overall information flow of each signaling pathway
      gg1 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE)
      gg2 <- rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE)
      print(gg1 + gg2)
  
  }
dev.off()

```

#FOR FIGURES 6B & C, 7E and S9A
```{r individual_cellchat}
groups<-c("Untd_d0","Untd_INF","UVB_d0","UVB_INF")
chat_utd <- readRDS(paste0(save_within, groups[1], "/cellchat_",groups[1],".rds"))
chat_inf <- readRDS(paste0(save_within, groups[2], "/cellchat_",groups[2],".rds"))
chat_uvb <- readRDS(paste0(save_within, groups[3], "/cellchat_",groups[3],".rds"))
chat_ub_inf <- readRDS(paste0(save_within, groups[4], "/cellchat_",groups[4],".rds"))
# test<-data.frame(chat_inf@DB[["interaction"]])
# write.csv(test,paste0(save_within,"test.csv"))
chat_inf@netP$pathways
chat_ub_inf@netP$pathways
# netVisual_chord_cell(chat_inf, signaling = c("ICAM"),legend.pos.x = 8)
# netVisual_chord_cell(chat_ub_inf, signaling = c("ICAM"),legend.pos.x = 8)

group <- "UVB_INF"
uvb_list <- c("ITGAL-ITGB2", "SEMA3", "SPP1", "MAG","VCAM","CD86")
#######The following PDF will contain the above listed interactions###
#######FIGURE 6C Itgal_Itgb2 - Icam1/2: Page 1 of the PDF
#######FIGURE 7E +UVB; +L.donovani(i.vi.) Chemokine and complement contribution: Page 7 of the PDF
######################################################################
pdf(paste0(save_within, group,"/", group, "_chords.pdf"))
  for(signal in uvb_list){
    netVisual_chord_cell(chat_ub_inf, signaling = signal,legend.pos.x = 8)
  }
  pathways.show <- c("CXCL","CCL","COMPLEMENT") 
  netVisual_chord_gene(chat_ub_inf, legend.pos.x = 15, signaling = pathways.show)
dev.off()
#######FIGURE S9 +UVB; +L.donovani(i.vi.) Contribution of top 6 pathways: Page 1 of the PDF
######################################################################
gg <- netAnalysis_contribution(chat_ub_inf, signaling = uvb_list, font.size = 8)
ggsave(filename=paste0(save_within, "/",group,"/", group, "_L-R_contribution.pdf"), plot=gg, width = 3, height = 2, units = 'in', dpi = 300)
#######FIGURE S9 +UVB; +L.donovani(i.vi.) Contribution of complement, CCLs and CXCLs: Page 1 of the PDF
######################################################################
cc <- netAnalysis_contribution(chat_ub_inf, signaling = pathways.show, font.size = 8)
ggsave(filename=paste0(save_within, "/",group,"/", group, "CCs_L-R_contribution.pdf"), plot=cc, width = 3, height = 2, units = 'in', dpi = 300)

#########################################################################
#########################################################################
group <- "Untd_INF"
inf_list <- c("CD34","GALECTIN","SELPLG","PARs","LIFR","CDH")
#######The following PDF will contain the above listed interactions###
#######FIGURE 6B CD34-Selp : Page 1 of the PDF
#######FIGURE 6B Selplg-Selp: Page 3 of the PDF
#######FIGURE 7E -UVB; +L.donovani(i.vi.) Chemokine and complement contribution: Page 7 of the PDF
######################################################################
pdf(paste0(save_within, group, "/", group,"_chords.pdf"))
  for(signal in inf_list){
    netVisual_chord_cell(chat_inf, signaling = signal,legend.pos.x = 8)
  }
  pathways.show <- c("CXCL","CCL","COMPLEMENT") 
  netVisual_chord_gene(chat_inf, legend.pos.x = 15, signaling = pathways.show)
dev.off()
#######FIGURE S9 -UVB; +L.donovani(i.vi.) Contribution of top 6 pathways: Page 1 of the PDF
######################################################################
gg <- netAnalysis_contribution(chat_inf, signaling = inf_list, font.size = 8)
ggsave(filename=paste0(save_within, "/",group,"/", group, "_L-R_contribution.pdf"), plot=gg, width = 3, height = 2, units = 'in', dpi = 300)
#######FIGURE S9 -UVB; +L.donovani(i.vi.) Contribution of complement, CCLs and CXCLs: Page 1 of the PDF
######################################################################
cc <- netAnalysis_contribution(chat_inf, signaling = pathways.show, font.size = 8)
ggsave(filename=paste0(save_within, "/",group,"/", group, "_CCs_L-R_contribution.pdf"), plot=cc, width = 3, height = 2, units = 'in', dpi = 300)


```



RUN THE FOLLOWING ONLY WHEN RE_RUNNING CELLCHAT
```{r load_source_data}
# load example dataset
mm2108 <- readRDS(paste0(exp_path, "integrated_mm2108_dims15_res0.4.rds"))
mm2108 <- SetIdent(mm2108, value = mm2108$CellTypeM)

```

Define all functions
```{r cellchat functions, echo=FALSE}
#Create CellChat Object
fn_create_cellchat <- function(data, meta, label) {
  cellchat <- createCellChat(object = data.input, meta = meta, group.by = "CellTypeM")
  #set database
  CellChatDB <- CellChatDB.mouse 
  showDatabaseCategory(CellChatDB)
  dplyr::glimpse(CellChatDB$interaction)
  
  # use a subset of CellChatDB for cell-cell communication analysis
  CellChatDB.use <- subsetDB(CellChatDB, search = c("Secreted Signaling","Cell-Cell Contact")) 
  # use all CellChatDB for cell-cell communication analysis
  #CellChatDB.use <- CellChatDB # simply use the default CellChatDB
  
  # set the used database in the object
  cellchat@DB <- CellChatDB.use
  
  return(cellchat)
}

#function to preprocess the object
fn_preprocess_cellchat <- function(obj) {
  # subset the expression data of signaling genes for saving computation cost
  obj <- subsetData(obj) # This step is necessary even if using the whole database
  #future::plan("multiprocess", workers = 4) # do parallel
  # Warning message:
  # Strategy 'multiprocess' is deprecated in future (>= 1.20.0). Instead, explicitly specify either 'multisession' or 'multicore'. In the current R session, 'multiprocess' equals 'multisession'. 
  #> parallelly::supportsMulticore
  obj <- identifyOverExpressedGenes(obj)
  obj <- identifyOverExpressedInteractions(obj)
  
  # project gene expression data onto PPI (Optional: when running it, USER should set `raw.use = FALSE` in the function `computeCommunProb()` in order to use the projected data)
  obj <- projectData(obj, PPI.mouse)
  return(obj)
}

#function
fn_infer_cell_cell <- function(obj){
  obj <- computeCommunProb(obj)
  # Filter out the cell-cell communication if there are only few number of cells in certain cell groups
  obj <- filterCommunication(obj, min.cells = 10)
  #compute probabilities
  obj <- computeCommunProbPathway(obj)
  #obtain aggregated signal
  obj <- aggregateNet(obj)
  ##compute centrality
  try(obj <- netAnalysis_computeCentrality(obj, slot.name = "netP"))
  
  return(obj)
}

# fn_auto_save_all_interactions <- function(obj, group, path){
#   pathways.show.all <- obj@netP$pathways
#   # check the order of cell identity to set suitable vertex.receiver
#   levels(obj@idents)
#   vertex.receiver = seq(1,4)
#   for (i in 1:length(pathways.show.all)) {
#       # Visualize communication network associated with both signaling pathway and individual L-R pairs
#       #nvh_gg <- netVisual(obj, signaling = pathways.show.all[i], vertex.receiver = vertex.receiver, layout = "hierarchy")
#       #ggsave(filename=paste0(path, group, pathways.show.all[i], "_L-R_hierarchy.png"), plot=nvh_gg, width = 3, height = 2, units = 'in', dpi = 300)
#       # Compute and visualize the contribution of each ligand-receptor pair to the overall signaling pathway
#       gg <- netAnalysis_contribution(obj, signaling = pathways.show.all[i], font.size = 8)
#       ggsave(filename=paste0(path, group, "_", pathways.show.all[i], "_L-R_contribution.pdf"), plot=gg, width = 3, height = 2, units = 'in', dpi = 300)
#   }
# }
# 
# fn_bubble_chord <- function(obj, group, path){
#     targets_use <- c(1:16)
#     # Bubble plot
#     # We can also show all the significant interactions (L-R pairs) from some cell groups to other cell groups using netVisual_bubble.
#     
#     # show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
#    list_cells_in_order <- c("FIV","FV","FIII","FI","Mac","T","FVI","EndoI","FII","Basal","tDP","RBCs","LEC_LV","Myofibroblasts","Schwann","Sebocytes")
#    cell_counter <- 1
#    for(cell in list_cells_in_order){
#     nv_gg<-netVisual_bubble(obj, sources.use = cell_counter, targets.use = targets_use, remove.isolate = FALSE, title.name = cell, thresh = 0.01)
#     ggsave(filename=paste0(path, cell  ,"_bubble_significant_LRs_inf_utd.pdf"), plot=nv_gg, width = 8, height = 10, units = 'in', dpi = 300)
#     cell_counter <- cell_counter + 1     
#    }
#    # nv_gg<-netVisual_bubble(obj, sources.use = 2, targets.use = targets_use, remove.isolate = FALSE, title.name = "FV")
#    # ggsave(filename=paste0(path, group, "FV_bubble_significant_LRs_inf_utd.pdf"), plot=nv_gg, width = 8, height = 10, units = 'in', dpi = 300)
#    # nv_gg<-netVisual_bubble(obj, sources.use = 5, targets.use = targets_use, remove.isolate = FALSE, title.name = "Mac")
#    # ggsave(filename=paste0(path, group, "Mac_bubble_significant_LRs_inf_utd.pdf"), plot=nv_gg, width = 8, height = 10, units = 'in', dpi = 300)
#    # nv_gg<-netVisual_bubble(obj, sources.use = 6, targets.use = targets_use, remove.isolate = FALSE, title.name = "T")
#    # ggsave(filename=paste0(path, group, "T_bubble_significant_LRs_inf_utd.pdf"), plot=nv_gg, width = 8, height = 10, units = 'in', dpi = 300)
#    
#   
#   pdf(paste0(path, group, "chord_significant_LRs_inf_utd.pdf"), width = 12, height = 12) 
#     
#     gg_scatter <- netAnalysis_signalingRole_scatter(obj, title = names(obj), show.legend = FALSE, label.size = 5, dot.size = 5, font.size = 12, )
#     print(gg_scatter)
#     #> Comparing communications on a single object
#     #> # show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
#     # show all the interactions sending from FIII
#     #show signals from (sources.use) from FibroblastV (2), mac (5) and T cells (6)
#     #netVisual_chord_gene(obj, sources.use = targets_use, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "ALL")
#     netVisual_chord_gene(obj, sources.use = 1, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "FIV")
#     netVisual_chord_gene(obj, sources.use = 2, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "FV")
#     netVisual_chord_gene(obj, sources.use = 3, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "FI")
#     netVisual_chord_gene(obj, sources.use = 4, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "FIII")
#     netVisual_chord_gene(obj, sources.use = 5, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "Mac")
#     netVisual_chord_gene(obj, sources.use = 6, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "T")
#     netVisual_chord_gene(obj, sources.use = 7, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "FVI")
#     netVisual_chord_gene(obj, sources.use = 8, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "Endo")
#     netVisual_chord_gene(obj, sources.use = 9, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "FII")
#     netVisual_chord_gene(obj, sources.use = 10, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "Basal")
#     netVisual_chord_gene(obj, sources.use = 11, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "tDP")
#     netVisual_chord_gene(obj, sources.use = 12, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "RBCs")
#     netVisual_chord_gene(obj, sources.use = 13, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "LEC_LV")
#     netVisual_chord_gene(obj, sources.use = 14, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "Myofibroblasts")
#     netVisual_chord_gene(obj, sources.use = 15, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "Schwann")
#     try(netVisual_chord_gene(obj, sources.use = 16, targets.use = targets_use, lab.cex = 0.75,legend.pos.y = 30, title.name = "Sebocytes"))
#     
#   dev.off()
# }
# 
# #create allulvials - pre select number of patterns 
# fn_alluvials <- function(obj, group, path){
#   pdf(paste0(path, group, "cell_communication_patterns_outgoing.pdf"), width = 8, height = 6)
#     # Here we run selectK to infer the number of patterns.
#   selectK(obj, pattern = "outgoing")
#   
#   # Both Cophenetic and Silhouette values begin to drop suddenly when the number of outgoing patterns is 3.
#     nPatterns = 3
#     obj <- identifyCommunicationPatterns(obj, pattern = "outgoing", k = nPatterns)
#   dev.off()
#     # river plot
#     river_gg<-netAnalysis_river(obj, pattern = "outgoing")
#     ggsave(filename=paste0(path, group, "_outgoing_river.pdf"), plot=river_gg, width = 5, height = 4, units = 'in', dpi = 300)
#     #> Please make sure you have load `library(ggalluvial)` when running this function
# 
#   pdf(paste0(path, group, "cell_communication_patterns_incoming.pdf"), width = 8, height = 6)
#   # Here we run selectK to infer the number of patterns.
#   selectK(obj, pattern = "incoming")
#   
#     nPatterns = 3
#     obj <- identifyCommunicationPatterns(obj, pattern = "incoming", k = nPatterns)
# 
#   dev.off()
#       # river plot
#     river_gg<-netAnalysis_river(obj, pattern = "incoming")
#     ggsave(filename=paste0(path, group, "_incoming_river.pdf"), plot=river_gg, width = 5, height = 4, units = 'in', dpi = 300)
#     #> Please make sure you have load `library(ggalluvial)` when running this function
# }

```
Run cellchat for individual datasets
```{r Run_cellchat_single}
#dataset to use
#table(mm2108.small$D_UVB)
# data <- mm2108#.small
# #rm(mm2108.small)
# rm(mm2108)
#intialise counter
count <- 1
for(group in unique(mm2108$D_UVB)){
  data.input = mm2108[["SCT"]]@data # SCT data matrix
  meta = mm2108@meta.data # a dataframe with rownames containing cell meta data
  cell.use = rownames(meta)[meta$D_UVB == group] # extract the cell names from disease data
  #reformat group name
  group_reformatted <- str_replace(group, "/", "_")
  #define save_path
  save_path <- paste0(save_within, group_reformatted, "/")
  # Prepare input data for CelChat analysis
  data.input = data.input[, cell.use]
  meta = meta[cell.use, ]

  unique(meta$CellTypeM) # check the cell labels
  #create cellchat object
  cellchat <- fn_create_cellchat(data.input, meta, "CellTypeM")
  #preprocess
  cellchat <- fn_preprocess_cellchat(cellchat)
  #infer
  cellchat <- fn_infer_cell_cell(cellchat)
  #save all interaction
  # fn_auto_save_all_interactions(cellchat, group_reformatted, save_path)
  # #save bubble/chord
  # fn_bubble_chord(cellchat, group_reformatted, save_path)
  # #create alluvials
  # fn_alluvials(cellchat, group_reformatted, save_path)
  #save cellchat object
  saveRDS(cellchat, paste0(save_path,"cellchat_",group_reformatted,".rds"))
  #update counter
  count <- count + 1
  #remove cellchat
  rm(cellchat)
}
   
rm(data)

```
